src/lib/userDatabase.js

```diff
@@ -0,0 +1,106 @@
+const {error} = require("lib/log");
+
+const userDatabase = {
+  _database: null,
+
+  init(keys = {"blockedLinks": []}, db = window.indexedDB) {
+    this._keys = keys;
+    return new Promise((resolve, reject) => {
+      const request = db.open("NewTabData", 1);
+
+      request.onerror = event => {
+        const errorString = event.target.errorCode + ": Cannot open an indexedDB connection";
+        reject(new Error(errorString));
+      };
+      request.onsuccess = event => {
+        // Save the database connection.
+        userDatabase._database = event.target.result;
+        resolve();
+      };
+      request.onupgradeneeded = event => {
+        // Note: After a successful upgrade, onsuccess will be triggered.
+        userDatabase._onDatabaseUpgrade(event);
+      };
+    });
+  },
+
+  save(objectStoreToWrite, objectStoreType, data, mockObjectStore) {
+    return new Promise((resolve, reject) => {
+      const transaction = userDatabase._database.transaction([objectStoreToWrite], "readwrite");
+      const objectStore = mockObjectStore || transaction.objectStore(objectStoreToWrite);
+
+      const request = objectStore.get(objectStoreType);
+      request.onsuccess = () => {
+        userDatabase._onWriteFetchRequestSuccess(request, data, objectStore, objectStoreType).then(resolve, reject);
+      };
+      request.onerror = event => {
+        const errorString = `${event.target.errorCode}: Failed to store object of type ${objectStoreType}`;
+        error(errorString);
+        reject(new Error(errorString));
+      };
+    });
+  },
+
+  load(objectStoreToRead, objectStoreType) {
+    const transaction = userDatabase._database.transaction([objectStoreToRead]);
+    const objectStore = transaction.objectStore(objectStoreToRead);
+    const request = objectStore.get(objectStoreType);
+    const transactionDescription = "Load data " + objectStoreType;
+    return userDatabase._setSimpleRequestHandlers(request, transactionDescription);
+  },
+
+  _setSimpleRequestHandlers(request, logString) {
+    return new Promise((resolve, reject) => {
+      request.onerror = event => {
+        const errorString = event.target.errorCode + ": " + logString;
```

Comment:
I'm not sure the error code buys us anything. In the IDB spec, it states that:

> "IndexedDB uses the following new DOMException types with their various messages. All of these new types will have a code value of 0 zero". 

Hence, the message is the only valuable bit of information. The code number is ambiguous at best, and kinda useless at worst.